### PR TITLE
210 backend trpc websocket integration websocket go brrr brrr

### PR DESCRIPTION
**Pull request by thienguen**
# What type of PR is this? (Check all that apply)

- [x] ✨**feature**: Introduces completely new code or new features.
- [x] 🐛**fix**: Implements changes that fix a bug. Ideally, reference an issue if present.
- [ ] ♻️**refactor**: Includes any code-related change that is neither a fix nor a feature.
- [x] ✅**build**: Encompasses all changes related to the build of the software, including changes to dependencies or the addition of new ones.
- [ ] ⚡️**test**: Pertains to all changes regarding tests, whether adding new tests or modifying existing ones.
- [ ] 🚰**ci**: Involves all changes related to the configuration of continuous integration, such as GitHub Actions or other CI systems.
- [ ] 📚**docs**: Includes all changes to documentation, such as README files, or any other documentation present in the repository.
- [ ] 🗑️**chore**: Captures all changes to the repository that do not fit into the above categories.

# Description

What did you change? How did you change it?

1. I got cooked in the backend, pretty pretty pretty rough
2. integrate trpc websocket in our db, having it ready in localhost
3. (blocker) couldn't test in deployment, since this is still in local host, I will be hosting it on my personal account of this repository (fork) and then go from there
4. All our procedures currently reside in https://trpc.io/docs/subscriptions.
5. I will address this as a "advanced feature" in my next issue.
6. create 2 testing environment for websocket, however, you should be focusing on the mobile environment.
7. both API handler now resides in `api.tsx` respectively
8. fixed `pnpm run vitest`
9. fixed migration
10. fixed packages versions (all fixed see commits history)
11. Separate context instances in trpc.ts (now we know its either in nextjs or in expo again)

# Tests

How was this tested?

Run like normal, invoke both the website (we treat it as a server here) and invoke the mobile app. Follow the instruction with screenshots down below

- [ ] Unit tests
- [x] Integration tests
- [ ] E2E tests
- [x] Manual tests
- [ ] Tests were NOT needed
- [ ] Other (explain below)

# [Optional] Screenshots

image

+ Have at least 2 instances of virtual devices running (WE USING EXPO GO WITH THIS, but if you can open the app on 2 devices using development build, feel free to do so)

+ In the 1st terminal (the website / "server") you should see

image

+ I'm running 3 virtual devices, aka 3 connections has established
+ Navigate to `nav` --> post2

image

+ type through the keyboard, and you should see T is typein (this is hard code)

image

+ Submit something, should also show in different devices since you can assume this test environment is their dedicated "room"

image

# I CAN'T RECORD THE ENTIRE SCREEN SINCE THE DEVICE KEEP TAKING THE FOCUS.

https://github.com/UNLV-CS472-672/2024-S-GROUP3-Barbell/assets/94078395/0a5b8cb1-0240-47d6-9380-4c1b7135f64b

# Documentation

- [x] Added to README.me
- [ ] Separate document
- [ ] NO documentation needed

Link to external documentation:

# [Optional] Are there any post-deployment tasks we need to perform?

TBD, will write how to test after the break...but I won't be writing more, though I do have a couple of stuff I want to organize in our repo

Update: n/a (will brought up in our meeting)

Update: (after talking to marcos) I can't test on deployment unless this code is in main (so I can sync and make it a "production", organization so nothing I can do (in a short time)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -5,5 +5,8 @@
     }
   ],
   "typescript.tsdk": "node_modules\\typescript\\lib",
-  "discord.enabled": true
+  "discord.enabled": true,
```
Comment (thienguen) [inline .vscode/settings.json:8]: oh...gotta remove this

**Comment by thienguen** [comment]
# WILL WRITE TEST, BUT IT IS READY

File: apps/nextjs/src/app/layout.tsx [hunk 2]
```diff
@@ -23,24 +23,27 @@ export const dynamic = 'force-dynamic'
...
-    description: 'Simple monorepo with shared backend for web & mobile apps',
-    url: 'https://create-t3-turbo.vercel.app',
-    siteName: 'Barbell',
-  },
-  twitter: {
-    card: 'summary_large_image',
-    site: '@barbell',
-    creator: '@barbell',
-  },
+  // openGraph: {
```
Comment (thienguen) [inline apps/nextjs/src/app/layout.tsx:26]: remove warning when running nextjs about not handle graph or twitter

File: packages/api/src/__tests__/post.router.test.ts [hunk 3]
```diff
@@ -1,33 +1,31 @@
-import type { inferProcedureInput } from '@trpc/server'
+import { describe, expect, it } from 'vitest'
 
-import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
-
-import { createCaller, RouterInputs, RouterOutputs } from '../..'
-import { AppRouter } from '../root'
+import { createCaller, RouterInputs } from '../..'
 import { createContextInner } from '../trpc'
```
Comment (thienguen) [inline packages/api/src/__tests__/post.router.test.ts:5]: this fixed testing after dropping title in Post...

File: packages/api/src/router/post.router.ts [hunk 4]
```diff
@@ -22,17 +22,20 @@ export const postRouter = createTRPCRouter({
   create: publicProcedure
     .input(
       z.object({
-        title: z.string().min(1),
+        // title: z.string().min(1),
         content: z.string().min(1),
+        authorId: z.number().int(),
       }),
     )
```
Comment (thienguen) [inline packages/api/src/router/post.router.ts:29]: back to create, instead of upsert.

File: packages/api/src/router/post2.router.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  isTypingUpdate: () => void
+}
+declare interface MyEventEmitter {
+  on<TEv extends keyof MyEvents>(event: TEv, listener: MyEvents[TEv]): this
+  off<TEv extends keyof MyEvents>(event: TEv, listener: MyEvents[TEv]): this
+  once<TEv extends keyof MyEvents>(event: TEv, listener: MyEvents[TEv]): this
+  emit<TEv extends keyof MyEvents>(event: TEv, ...args: Parameters<MyEvents[TEv]>): boolean
+}
+
+class MyEventEmitter extends EventEmitter {}
```
Comment (thienguen) [inline packages/api/src/router/post2.router.ts:26]: Websocket logic

File: packages/db/package.json [hunk 6]
```diff
@@ -18,7 +18,7 @@
     "lint": "eslint .",
```
Comment (thienguen) [inline packages/db/package.json:18]: gotta do migration more often

File: packages/db/prisma/schema.prisma [hunk 7]
```diff
@@ -67,6 +67,23 @@ enum Gender {
   PREFERNOTTOSAY
 }
```
Comment (thienguen) [inline packages/db/prisma/schema.prisma:69]: this is for testing environment, will show lata

File: packages/db/src/new-gen-data/post.json [hunk 8]
```diff
@@ -1,104 +1,95 @@
 [
   {
```
Comment (thienguen) [inline packages/db/src/new-gen-data/post.json:2]: remove id, creation completed

**Review by thienguen — COMMENTED** [review]
self-reviewed.

There is no example can be found on the internet, and the use of AI couldn't help anything in this PR.

**Review by mva919 — APPROVED** [review]
Working on my end

**Pull request merged**